cmd/kubeadm/app/util/config/masterconfig.go

```diff
@@ -61,47 +83,58 @@ func SetInitDynamicDefaults(cfg *kubeadmapi.InitConfiguration) error {
 		if err != nil {
 			return err
 		}
-		cfg.BootstrapTokens[i].Token = token
+		(*cfg)[i].Token = token
 	}
 
+	return nil
+}
+
+// SetNodeRegistrationDynamicDefaults checks and sets configuration values for the NodeRegistration object
+func SetNodeRegistrationDynamicDefaults(cfg *kubeadmapi.NodeRegistrationOptions, masterTaint bool) error {
 	var err error
-	cfg.NodeRegistration.Name, err = nodeutil.GetHostname(cfg.NodeRegistration.Name)
+	cfg.Name, err = nodeutil.GetHostname(cfg.Name)
 	if err != nil {
 		return err
 	}
 
 	// Only if the slice is nil, we should append the master taint. This allows the user to specify an empty slice for no default master taint
-	if cfg.NodeRegistration.Taints == nil {
-		cfg.NodeRegistration.Taints = []v1.Taint{kubeadmconstants.MasterTaint}
+	if masterTaint && cfg.Taints == nil {
+		cfg.Taints = []v1.Taint{kubeadmconstants.MasterTaint}
 	}
 
-	// Do all the defaulting for the nested ClusterConfiguration as well
-	return SetClusterDynamicDefaults(&cfg.ClusterConfiguration)
+	return nil
 }
 
-// SetClusterDynamicDefaults checks and sets configuration values for the InitConfiguration object
-func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration) error {
-	// Default all the embedded ComponentConfig structs
-	componentconfigs.Known.Default(cfg)
-
+// SetAPIEndpointDynamicDefaults checks and sets configuration values for the APIEndpoint object
+func SetAPIEndpointDynamicDefaults(cfg *kubeadmapi.APIEndpoint) error {
 	// validate cfg.API.AdvertiseAddress.
-	addressIP := net.ParseIP(cfg.API.AdvertiseAddress)
-	if addressIP == nil && cfg.API.AdvertiseAddress != "" {
-		return fmt.Errorf("couldn't use \"%s\" as \"apiserver-advertise-address\", must be ipv4 or ipv6 address", cfg.API.AdvertiseAddress)
+	addressIP := net.ParseIP(cfg.AdvertiseAddress)
+	if addressIP == nil && cfg.AdvertiseAddress != "" {
+		return fmt.Errorf("couldn't use \"%s\" as \"apiserver-advertise-address\", must be ipv4 or ipv6 address", cfg.AdvertiseAddress)
 	}
 	// Choose the right address for the API Server to advertise. If the advertise address is localhost or 0.0.0.0, the default interface's IP address is used
 	// This is the same logic as the API Server uses
 	ip, err := netutil.ChooseBindAddress(addressIP)
 	if err != nil {
 		return err
 	}
-	cfg.API.AdvertiseAddress = ip.String()
-	ip = net.ParseIP(cfg.API.AdvertiseAddress)
+	cfg.AdvertiseAddress = ip.String()
+
+	return nil
+}
+
+// SetClusterDynamicDefaults checks and sets configuration values for the InitConfiguration object
+func SetClusterDynamicDefaults(cfg *kubeadmapi.ClusterConfiguration, advertiseAddress string) error {
+	// Default all the embedded ComponentConfig structs
+	componentconfigs.Known.Default(cfg)
+
+	ip := net.ParseIP(advertiseAddress)
```

Comment:
this will probably conflict with:
https://github.com/kubernetes/kubernetes/pull/67397
either this or that PR will need a rebase.